src/main/java/org/traccar/database/DataManager.java

```diff
@@ -328,6 +328,18 @@ public void updateDeviceStatus(Device device) throws SQLException {
                 .executeQuery(Position.class);
     }
 
+    public Position getPrecedingPosition(long deviceId, Date date) throws SQLException {
+        Collection<Position> positions = QueryBuilder.create(dataSource, getQuery("database.selectPrecedingPosition"))
+                .setLong("deviceId", deviceId)
+                .setDate("time", date)
+                .executeQuery(Position.class);
+        if (positions.isEmpty()) {
+            return null;
+        } else {
+            return positions.iterator().next();
+        }
```

Comment:
Sorry, just noticed another thing. Can we use `executeQuerySingle` here instead?